google/resource_dataproc_job.go

```diff
@@ -0,0 +1,840 @@
+	}
+	return e
+}
+
+func resourceDataprocJobRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	region := d.Get("region").(string)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	job, err := config.clientDataproc.Projects.Regions.Jobs.Get(
+		project, region, d.Id()).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Dataproc Job %q", d.Id()))
+	}
+
+	d.Set("labels", job.Labels)
+	d.Set("cluster", job.Placement.ClusterName)
+	d.Set("status", job.Status.State)
+	d.Set("outputUri", job.DriverOutputResourceUri)
+
+	if job.PysparkJob != nil {
+		d.Set("pyspark_config", getTfPySparkConfig(job.PysparkJob))
+	}
+	if job.SparkJob != nil {
+		d.Set("spark_config", getTfSparkConfig(job.SparkJob))
+	}
+	if job.HadoopJob != nil {
+		d.Set("hadoop_config", getHadoopConfig(job.HadoopJob))
+	}
+	if job.HiveJob != nil {
+		d.Set("hive_config", getHiveConfig(job.HiveJob))
+	}
+	if job.PigJob != nil {
+		d.Set("pig_config", getPigConfig(job.PigJob))
+	}
+	if job.SparkSqlJob != nil {
+		d.Set("sparksql_config", getSparkSqlConfig(job.SparkSqlJob))
+	}
+	return nil
+}
+
+func resourceDataprocJobDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region := d.Get("region").(string)
+	forceDelete := d.Get("force_delete").(bool)
+	timeoutInMinutes := int(d.Timeout(schema.TimeoutDelete).Minutes())
+
+	if forceDelete {
+		log.Printf("[DEBUG] Attempting to first cancel Dataproc job %s if its still running ...", d.Id())
+
+		_, _ = config.clientDataproc.Projects.Regions.Jobs.Cancel(
+			project, region, d.Id(), &dataproc.CancelJobRequest{}).Do()
+		// ignore error if we get one - job may be finished already and not need to
+		// be cancelled. We do however wait for the state to be one that is
+		// at least not active
+		waitErr := dataprocJobOperationWait(config, region, project, d.Id(),
+			"Cancelling Dataproc job", timeoutInMinutes, 1)
+		if waitErr != nil {
+			return waitErr
+		}
+
+	}
+
+	log.Printf("[DEBUG] Deleting Dataproc job %s", d.Id())
+	_, err = config.clientDataproc.Projects.Regions.Jobs.Delete(
+		project, region, d.Id()).Do()
+	if err != nil {
+		return err
+	}
+
+	waitErr := dataprocDeleteOperationWait(config, region, project, d.Id(),
+		"Deleting Dataproc job", timeoutInMinutes, 1)
+	if waitErr != nil {
+		return waitErr
+	}
+
+	log.Printf("[INFO] Dataproc job %s has been deleted", d.Id())
+	d.SetId("")
+
+	return nil
+}
+
+// ---- PySpark Job ----
+
+func pySparkTFSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:          schema.TypeList,
+		Optional:      true,
+		ForceNew:      true,
+		MaxItems:      1,
+		ConflictsWith: []string{"spark_config", "hadoop_config", "hive_config", "pig_config", "sparksql_config"},
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+
+				"main_python_file": {
```

Comment:
if it's not too much effort, let's try to match the API names of these fields (main_python_file_uri, file_uri, etc)